lms/validation/_api.py

```diff
@@ -0,0 +1,40 @@
+"""Schema for JSON APIs exposed to the frontend."""
+
+from webargs import fields
+
+from lms.validation._helpers import PyramidRequestSchema
+
+
+__all__ = ["ApiRecordSubmissionSchema"]
+
+
+class ApiRecordSubmissionSchema(PyramidRequestSchema):
+    """Schema for validating requests from the frontend to record submissions."""
+
+    locations = ["json"]
+
+    document_url = fields.Str()
+    """
+    URL of the document for this assignment.
+    """
+
+    canvas_file_id = fields.Str()
+    """
+    ID of the Canvas file for this assignment.
+    """
+
+    h_username = fields.Str(required=True)
+    """
+    h username generated for the active user.
+    """
```

Comment:
> Or alternatively the h_username property could be moved out of LTILaunchResource and into some other class, since it's now also needed outside of LTI launches.

I think that would be better, as it seems to me that things could be added to `LTILaunchResource` in future which are appropriate for LTI launches but not for other contexts.